Add a STM32 Digital Temperature Sensor and add it to STM32H7 family DTS

aurel32 opened the pull request:
Tested on a Nucleo h7a3zi board, it gives similar results as the analog temperature sensor.

aurel32 committed:
drivers: sensor: st: add driver for the STM32 Digital Temperature Sensor

This add basic support for the STM32 Digital Temperature Sensor found
notably on the STM32H7 series. It work in interrupt mode and support
basic power device management.

It does not support the more advanced features like using the
temperature threshold, triggers from LPTIM or using the LSE clock in
during sleep or stop.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
dts: arm: st: h723/h7a3: add digi_dietemp node into DTSI file

Add a digi_dietemp node for the STM32 Digital Temperature Sensor into
stm32h723.dtsi (used as a base for H723, H725, H730 and H735) and
stm32h7a3.dtsi (used as a base for H7A3, H7B0 and H7B3).

The sensor is not available on other H7 SoCs.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

ubieda commented on drivers/sensor/st/Kconfig:
```diff
@@ -32,6 +32,7 @@ source "drivers/sensor/st/lsm6dsv16x/Kconfig"
 source "drivers/sensor/st/lsm9ds0_gyro/Kconfig"
 source "drivers/sensor/st/lsm9ds0_mfd/Kconfig"
 source "drivers/sensor/st/qdec_stm32/Kconfig"
+source "drivers/sensor/st/stm32_dts/Kconfig"
```
May we use a different acronym to describe digital temperature sensor? e.g: dtemp or similar.

aurel32 replied:
Yes, it's possible to rename it into `stm32_dtemp` if people feel it is better like that.

erwango replied:
I agree we should absolutely avoid dts :)
`stm32_dtemp` could be ok we think there won't be any confusion with die temp (for instance if some other vendor comes up with a vnd_dtemp for his own die temp sensor).
`stm32_digi_temp` is less synthetic but should avoid possible confusion.

ubieda commented on drivers/sensor/st/stm32_dts/stm32_dts.c:
```diff
@@ -0,0 +1,283 @@
+	}
+
+	/* Trigger a measurement */
+	cfg->base->CFGR1 |= DTS_CFGR1_TS1_START;
```
Is there a HAL that could abstract all these direct register operations away from this sensor driver?

aurel32 replied:
There is a HAL, but that will make the driver way more complicated, and things like power management would be more difficult to implement.

OTOH, I wish there is a LL like for other drivers, but it is unfortunately not the case. I guess the driver can be converted later if/when it get added to stm32cube.

aurel32 replied:
Another drawback of the HAL, is that it only returns an integer temperature.

aurel32 replied:
I have just opened https://github.com/STMicroelectronics/stm32h7xx_hal_driver/issues/56

ubieda commented on dts/bindings/sensor/st,stm32-dts.yaml:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2024, Aurelien Jarno
+# SPDX-License-Identifier: Apache-2.0
+
+description: STM32 family Digital Temperature Sensor node
```
Right now there's stm32_temp driver in-tree (based on ADC readings).

Is there a way we could abstract to the user which implementation is selected? I'm thinking of Kconfig selection based on the presence of this DTS node. That way, the user doesn't really has to _choose_ between the two: if it's digital use that one.

aurel32 replied:
I have mixed feeling about that. Users might have reasons to choose about that. While it is a simpler driver than the analog one with the ADC, it is less accurate and use ~10 times more power (excluding the ADC consumption, but users might have to use the ADC anyway).

aurel32 replied:
> Right now there's stm32_temp driver in-tree (based on ADC readings).
> 
> Is there a way we could abstract to the user which implementation is selected? I'm thinking of Kconfig selection based on the presence of this DTS node. That way, the user doesn't really has to _choose_ between the two: if it's digital use that one.

The only way I can see to abstract that given they are two different peripherals with different device tree properties is to remove the analog sensor from the device tree for the STM32 devices having a digital sensor, and point the `die_temp` label to it. This assumes users are referring to it via label and not via name or compat.

erwango replied:
> Users might have reasons to choose about that. While it is a simpler driver than the analog one with the ADC, it is less accurate and use ~10 times more power (excluding the ADC consumption, but users might have to use the ADC anyway).

That would be worth documenting in this description section. Once the 2 bindings will be available, they'll need element to help in deciding which one to choose.

erwango commented on drivers/sensor/st/stm32_dts/stm32_dts.c:
```diff
@@ -0,0 +1,283 @@
+LOG_MODULE_REGISTER(stm32_dts, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Constants */
+#define ONE_MHZ			1000000 /* Hz */
```
Not used

erwango reviewed (changes requested):
Thanks for this addition @aurel32.
To be complete, we'd need a way to at least build this driver in CI (using `samples/sensor/die_temp_polling` ?)

kartben commented on drivers/sensor/st/stm32_dts/stm32_dts.c:
```diff
@@ -0,0 +1,283 @@
+	uint16_t raw;
+};
+
+struct stm32_dts_config {
```
Looks like there's several other "DTS" occurrences that could use renaming

decsny commented on dts/bindings/sensor/st,stm32-dts.yaml:
```diff
@@ -0,0 +1,18 @@
+
+description: STM32 family Digital Temperature Sensor node
+
+compatible: "st,stm32-dts"
```
really do not like this compatible, because of the "dts"

aurel32 committed:
samples: sensors: nucleo_h7a3zi_q board added to die_temp_polling

Added board nucleo_h7a3zi_q overlay file to permit testing the STM32
Digital Temperature Sensor.

Also added the board as integration platform to the sample.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
I have just pushed a new version renaming the sensor to `stm32_digi_temp`, the device tree compat to `st,stm32-digi-temp` and the node label to `digi_die_temp`.

I have added a board overlay for the `die_temp_polling` sample. It outputs:

```
*** Booting Zephyr OS build v3.6.0-4233-gefac3ef4f2b8 ***
CPU Die temperature[digi_dietemp@58006800]: 30.4 °C
CPU Die temperature[digi_dietemp@58006800]: 30.1 °C
CPU Die temperature[digi_dietemp@58006800]: 30.1 °C

...
```

FRASTM commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -1067,6 +1067,15 @@
 			clocks = <&rcc STM32_CLOCK_BUS_AHB2 0x00000001>;
 			status = "disabled";
 		};
+
```
Not available for all the stm32h7, so be careful when adding the node there.
stm32h723x and stm32h7a3 series have but not the stm32h743 and next

aurel32 replied:
Good catch. I have just fixed that.

FRASTM reviewed (commented):
Thanks a lot for adding this new driver !
It's a interesting enhancement

FRASTM commented on drivers/sensor/st/stm32_digi_temp/stm32_digi_temp.c:
```diff
@@ -0,0 +1,283 @@
+	const struct stm32_digi_temp_config *cfg = dev->config;
+
+	/* Clear interrupt */
+	cfg->base->ICIFR |= DTS_ICIFR_TS1_CITEF;
```
Did you try using the stm32h7xx_hal_dts functions like __HAL_DTS_CLEAR_FLAG ?

aurel32 replied:
Nope I haven't tried that, I was really looking for a LL header, and therefore didn't look at that one in details. I will give it a try.

aurel32 replied:
@FRASTM This doesn't work as this macro takes a `DTS_HandleTypeDef` handle as parameter, which is the HAL specific object on which the `HAL_DTS_*` methods act.

aurel32 replied:
OTOH, I have just pushed a new version using `CLEAR_BIT`, `SET_BIT` and `MODIFY_REG`.

erwango reviewed (changes requested):
Few late comments that I forgot to commit, sorry.

ajarmouni-st commented on drivers/sensor/st/stm32_digi_temp/Kconfig:
```diff
@@ -0,0 +1,18 @@
+	  Enable the driver for STM32 digital temperature sensor. This sensor
+	  is different from the STM32 analog temperature sensor, which is read
+	  by an ADC. While both drivers have similar code fooprint, the analog
+	  temperature driver also require the ADC driver to be enabled. The
```
Also `code footprint`

aurel32 replied:
Good catch, both fixed.

aurel32 commented on drivers/sensor/st/stm32_digi_temp/stm32_digi_temp.c:
```diff
@@ -0,0 +1,283 @@
+
+	/* Init mutex and semaphore */
+	k_mutex_init(&data->mutex);
+	k_sem_init(&data->sem_isr, 0, K_SEM_MAX_LIMIT);
```
Thanks you are indeed correct. Note however that the vast majority of the sensor drivers use `K_SEM_MAX_LIMIT`.

erwango replied:
Indeed, this is worrisome. I guess this pure copy/paste of an initial mistake. Though, there's always a risk when fixing this kind of things.

aurel32 commented on drivers/sensor/st/stm32_digi_temp/stm32_digi_temp.c:
```diff
@@ -0,0 +1,283 @@
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	/* Wait for the sensor to be ready (~40µS delay after enabling it) */
+	while ((cfg->base->SR & DTS_SR_TS1_RDY) == 0) {
```
I have just implemented that. This makes the driver 16 bytes smaller.

erwango reviewed (changes requested):
Thanks for the changes!

One very last request and this will be fine to me:
Unless I'm wrong, you'll need to add `nucleo_h7a3zi_q` as integration platform to the sample to ensure it is taken into account in test campaigns.

(We'll also need to tune this later on to get it ran on our test bench, but we'll take care of it).

aurel32 commented:
> One very last request and this will be fine to me: Unless I'm wrong, you'll need to add `nucleo_h7a3zi_q` as integration platform to the sample to ensure it is taken into account in test campaigns.

That should be done now.

erwango reviewed (dismissed):
Thanks!

ubieda commented on drivers/sensor/st/stm32_digi_temp/stm32_digi_temp.c:
```diff
@@ -0,0 +1,286 @@
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	/* Wait for the sensor to be ready (~40µS delay after enabling it) */
+	while (READ_BIT(dts->SR, DTS_SR_TS1_RDY) == 0) {
+		k_yield();
+	}
+
+	/* Trigger a measurement */
+	SET_BIT(dts->CFGR1, DTS_CFGR1_TS1_START);
+	CLEAR_BIT(dts->CFGR1, DTS_CFGR1_TS1_START);
+
+	/* Wait for interrupt */
+	k_sem_take(&data->sem_isr, K_FOREVER);
```
Non-blocker suggestion: should we have a fail-safe on the `while-loop` and `k_sem_take()`? I'm thinking of returning `-ETIMEDOUT` or similar in case the happy-path does not work.

aurel32 replied:
Do we really need to do that with `K_FOREVER`. Looking at the documentation, it could return:
- `0` – Semaphore taken.
- `-EBUSY` – Returned without waiting.
- `-EAGAIN` – Waiting period timed out, or the semaphore was reset during the waiting period.

It seems that the `-EBUSY` case can only happens when using `K_NO_WAIT`. For `-EAGAIN`, the period timeout can't happen as `K_FOREVER` is used, and the semaphore reset also can't happened as it is only initialized in `stm32_digi_temp_init` as part of the driver initialization.

Therefore I don't think we need to add a while loop around it. In addition, I haven't found such a structure (with `K_FOREVER`)  elsewhere in the Zephyr's code.

ubieda replied:
Perhaps I didn't explain myself: 

WRT the while-loop: I'm referring to
``` c
	while (READ_BIT(dts->SR, DTS_SR_TS1_RDY) == 0) {
		k_yield();
	}
```

WRT the semaphore: I'm alluding of a scenario where the semaphore is not given, therefore the execution path is blockde. Handling such case involves changing the semaphore from K_FOREVER to a specific timeout and handling the error-codes you described.

MaureenHelm commented on drivers/sensor/st/stm32_digi_temp/stm32_digi_temp.c:
```diff
@@ -0,0 +1,286 @@
+	k_mutex_init(&data->mutex);
+	k_sem_init(&data->sem_isr, 0, 1);
+
+	/* Configure interrupt */
+	IRQ_CONNECT(
+		DT_INST_IRQN(0),
+		DT_INST_IRQ(0, priority),
+		stm32_digi_temp_isr,
+		DEVICE_DT_INST_GET(0),
+		0);
+	irq_enable(DT_INST_IRQN(0));
```
We shouldn't hard code an instance number within a function that takes a `struct device *` as an argument

MaureenHelm commented on drivers/sensor/st/stm32_digi_temp/stm32_digi_temp.c:
```diff
@@ -0,0 +1,286 @@
+
+static struct stm32_digi_temp_data stm32_digi_temp_dev_data;
+
+static struct stm32_digi_temp_config stm32_digi_temp_dev_config = {
```
This can be `const`
